使用系统自带相册选择图片会导致APP崩溃

## 问题描述 | Problem Description

某个版本之前是可以正常使用的，因为是不常用的备用机，所以忘记具体是哪个版本之前了，目前更新了最新版后出现了该问题。

具体情况：选择 “照片”按钮 来添加图片↓
如果选择系统自带的“相册” - 发生崩溃（APP闪退），重新打开APP后进入“安全模式”并弹出下面的崩溃报告；
如果选择系统自带的“文件管理器” - 添加失败，什么也不会发生，无提示；
如果选择第三方APP“质感文件” - 能成功添加图片。
但是“质感文件”无法预览图片，还要每次都手动去指定路径翻图片，体验并不好。

### 崩溃报告：
Thread: main
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=913791392, result=-1, data=Intent { dat=content://media/external/images/media/1556760 clip={(has clip)} }} to activity {me.rerere.rikkahub/me.rerere.rikkahub.RouteActivity}: java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaProvider uri content://media/external/images/media/1556760 from pid=22193, uid=10527 requires android.permission.READ_EXTERNAL_STORAGE, or grantUriPermission()
	at android.app.ActivityThread.deliverResults(ActivityThread.java:4774)
	at android.app.ActivityThread.handleSendResult(ActivityThread.java:4818)
	at android.app.ActivityThread.-wrap20(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1843)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:167)
	at android.app.ActivityThread.main(ActivityThread.java:7146)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:820)
Caused by: java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaProvider uri content://media/external/images/media/1556760 from pid=22193, uid=10527 requires android.permission.READ_EXTERNAL_STORAGE, or grantUriPermission()
	at android.os.Parcel.readException(Parcel.java:2013)
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
	at android.content.ContentProviderProxy.query(ContentProviderNative.java:418)
	at android.content.ContentResolver.query(ContentResolver.java:756)
	at android.content.ContentResolver.query(ContentResolver.java:705)
	at android.content.ContentResolver.query(ContentResolver.java:663)
	at me.rerere.rikkahub.data.files.FilesManager.getFileNameFromUri(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:20)
	at me.rerere.rikkahub.data.files.FilesManager.createChatFilesByContents(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:49)
	at me.rerere.rikkahub.ui.components.ai.ChatInputKt$$ExternalSyntheticLambda4.invoke(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:88)
	at androidx.activity.OnBackPressedDispatcher$OnBackPressedEventInput$$ExternalSyntheticLambda0.onActivityResult(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:11)
	at androidx.activity.ComponentActivity$activityResultRegistry$1.dispatchResult(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:49)
	at androidx.activity.ComponentActivity.onActivityResult(r8-map-id-4a70cd041f559e1e55fe4897eb83d2a5ecc50c3db4ef600a8a16aee2571b66cb:3)
	at android.app.Activity.dispatchActivityResult(Activity.java:7570)
	at android.app.ActivityThread.deliverResults(ActivityThread.java:4770)
	... 9 more


## 截图 | Screenshot

<img width="1080" height="1252" alt="Image" src="https://github.com/user-attachments/assets/95de0ce0-1f89-416d-b7a6-83067fe8ca24" />

<img width="1080" height="2102" alt="Image" src="https://github.com/user-attachments/assets/7653a4c4-e500-4868-9ccc-b57ce65b2803" />

## 版本信息 | Version Information

APP版本: v2.1.7

安卓版本: Android 8.1.0 (SDK 27)

手机型号: 坚果Pro2S

--- a/app/src/main/java/me/rerere/rikkahub/data/files/FilesManager.kt
+++ b/app/src/main/java/me/rerere/rikkahub/data/files/FilesManager.kt
@@ -139,9 +139,11 @@ class FilesManager(
                 if (!file.exists()) {
                     file.createNewFile()
                 }
-                context.contentResolver.openInputStream(uri)?.use { inputStream ->
-                    file.outputStream().use { outputStream ->
-                        inputStream.copyTo(outputStream)
+                val inputStream = context.contentResolver.openInputStream(uri)
+                    ?: error("Failed to open input stream for $uri")
+                inputStream.use { input ->
+                    file.outputStream().use { output ->
+                        input.copyTo(output)
                     }
                 }
                 val guessedMime = sourceMime ?: guessMimeType(file, sourceName)

--- a/app/src/main/java/me/rerere/rikkahub/data/files/FilesManager.kt
+++ b/app/src/main/java/me/rerere/rikkahub/data/files/FilesManager.kt
@@ -131,15 +131,14 @@ class FilesManager(
             dir.mkdirs()
         }
         uris.forEach { uri ->
-            val sourceName = getFileNameFromUri(uri) ?: uri.lastPathSegment ?: "file"
-            val sourceMime = getFileMimeType(uri)
-            val fileName = buildUuidFileName(displayName = sourceName, mimeType = sourceMime)
-            val file = dir.resolve(fileName)
-            if (!file.exists()) {
-                file.createNewFile()
-            }
-            val newUri = file.toUri()
             runCatching {
+                val sourceName = getFileNameFromUri(uri) ?: uri.lastPathSegment ?: "file"
+                val sourceMime = getFileMimeType(uri)
+                val fileName = buildUuidFileName(displayName = sourceName, mimeType = sourceMime)
+                val file = dir.resolve(fileName)
+                if (!file.exists()) {
+                    file.createNewFile()
+                }
                 context.contentResolver.openInputStream(uri)?.use { inputStream ->
                     file.outputStream().use { outputStream ->
                         inputStream.copyTo(outputStream)
@@ -147,7 +146,7 @@ class FilesManager(
                 }
                 val guessedMime = sourceMime ?: guessMimeType(file, sourceName)
                 trackUploadFile(file = file, displayName = sourceName, mimeType = guessedMime)
-                newUris.add(newUri)
+                newUris.add(file.toUri())
             }.onFailure {
                 it.printStackTrace()
                 Log.e(TAG, "createChatFilesByContents: Failed to save file from $uri", it)
@@ -425,31 +424,39 @@ class FilesManager(
     }
 
     fun getFileNameFromUri(uri: Uri): String? {
-        var fileName: String? = null
-        val projection = arrayOf(
-            OpenableColumns.DISPLAY_NAME,
-            DocumentsContract.Document.COLUMN_DISPLAY_NAME
-        )
-        context.contentResolver.query(uri, projection, null, null, null)?.use { cursor ->
-            if (cursor.moveToFirst()) {
-                val documentDisplayNameIndex =
-                    cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME)
-                if (documentDisplayNameIndex != -1) {
-                    fileName = cursor.getString(documentDisplayNameIndex)
-                } else {
-                    val openableDisplayNameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
-                    if (openableDisplayNameIndex != -1) {
-                        fileName = cursor.getString(openableDisplayNameIndex)
+        return runCatching {
+            var fileName: String? = null
+            val projection = arrayOf(
+                OpenableColumns.DISPLAY_NAME,
+                DocumentsContract.Document.COLUMN_DISPLAY_NAME
+            )
+            context.contentResolver.query(uri, projection, null, null, null)?.use { cursor ->
+                if (cursor.moveToFirst()) {
+                    val documentDisplayNameIndex =
+                        cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME)
+                    if (documentDisplayNameIndex != -1) {
+                        fileName = cursor.getString(documentDisplayNameIndex)
+                    } else {
+                        val openableDisplayNameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
+                        if (openableDisplayNameIndex != -1) {
+                            fileName = cursor.getString(openableDisplayNameIndex)
+                        }
                     }
                 }
             }
-        }
-        return fileName
+            fileName
+        }.onFailure {
+            Log.w(TAG, "getFileNameFromUri: Failed to query display name for $uri", it)
+        }.getOrNull()
     }
 
     fun getFileMimeType(uri: Uri): String? {
         return when (uri.scheme) {
-            "content" -> context.contentResolver.getType(uri)
+            "content" -> runCatching {
+                context.contentResolver.getType(uri)
+            }.onFailure {
+                Log.w(TAG, "getFileMimeType: Failed to resolve MIME for $uri", it)
+            }.getOrNull()
             else -> null
         }
     }

